Update compare_set implementation for FileStore and HashStore

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57203 Update compare_set docs
* #57191 Add compare_set to trampoline class, add typing and formatting
* **#57175 Update compare_set implementation for FileStore and HashStore**

Update other Store implementations to add the value when current value is empty to match the amendment made to TCPStore (#55636). Added test to cover this case.

Test:
`pytest -vs test/distributed/test_c10d_common.py -k compare_set`

Differential Revision: [D28069380](https://our.internmc.facebook.com/intern/diff/D28069380)

H-Huang committed:
Update compare_set implementation for FileStore and HashStore

[ghstack-poisoned]

H-Huang commented on torch/lib/c10d/FileStore.cpp:
```diff
@@ -140,7 +140,8 @@ class File {
 #ifdef _WIN32
-      fd_ = syscall(std::bind(::open, path.c_str(), flags | _O_BINARY, _S_IREAD | _S_IWRITE));
+      fd_ = syscall(std::bind(
+          ::open, path.c_str(), flags | _O_BINARY, _S_IREAD | _S_IWRITE));
```
This change and above are purely formatting changes

H-Huang committed:
Update on "Update compare_set implementation for FileStore and HashStore"


Update other Store implementations to add the value when current value is empty to match the amendment made to TCPStore (#55636). Added test to cover this case.

Test:
`pytest -vs test/distributed/test_c10d_common.py -k compare_set`

Differential Revision: [D28069380](https://our.internmc.facebook.com/intern/diff/D28069380)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/FileStore.cpp:
```diff
@@ -310,9 +311,15 @@ std::vector<uint8_t> FileStore::compareSet(
   pos_ = refresh(file, pos_, cache_);
   if (cache_.count(regKey) == 0) {
+    // if the key does not exist and current value is empty then set it
+    if (currentValue.empty()) {
```
Nit: Right now you have two places in the function where you write a new key/value pair to the file; one here at line 315 and the other at line 323.You can most likely rearrange the `if` conditions and consolidate them into one single place.

H-Huang replied:
Thanks! Updated

cbalioglu commented on torch/lib/c10d/FileStore.cpp:
```diff
@@ -310,9 +311,15 @@ std::vector<uint8_t> FileStore::compareSet(
   // it might be outdated
   pos_ = refresh(file, pos_, cache_);
   if (cache_.count(regKey) == 0) {
+    // if the key does not exist and current value is empty then set it
```
Nit: Maybe you should re-word this comment. If the key does not exist, then there can't be any "current value". What you mean here with "current value" is the value of the `currentValue` parameter.

H-Huang replied:
Updated the comment

cbalioglu commented on torch/lib/c10d/HashStore.cpp:
```diff
@@ -24,10 +24,14 @@ std::vector<uint8_t> HashStore::compareSet(
-  if (it == map_.end()){
+  if (it == map_.end()) {
+    if (currentValue.empty()) {
+      map_[key] = newValue;
```
Nit: Same comment as above for `FileStore`.

cbalioglu reviewed (commented):
Looks good to me. Just left a few minor comments. Let me know whether you want to address them; otherwise, I can approve it now.

H-Huang committed:
Update on "Update compare_set implementation for FileStore and HashStore"


Update other Store implementations to add the value when current value is empty to match the amendment made to TCPStore (#55636). Added test to cover this case.

Test:
`pytest -vs test/distributed/test_c10d_common.py -k compare_set`

Differential Revision: [D28069380](https://our.internmc.facebook.com/intern/diff/D28069380)

[ghstack-poisoned]

cbalioglu reviewed (approved):
LGTM
